### ISSUE TITLE
Conditional parameter type based on generic type has inconsistent behavior when generic type passed as generic argument 

### ISSUE THREAD

**Issue by Hookyns** [microsoft/TypeScript#46155]
# Bug Report

### 🔎 Search Terms
generic conditional parameter type inconsistent behavior

_I tried to search a lot of terms but I don't even know how to name this problem._

### 🕗 Version & Regression Information
- This is the behavior in every version I tried (I tried all playground versions), and I reviewed the FAQ.

### ⏯ Playground Link
[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.0-dev.20210930#code/MYGwhgzhAEDKD2BbApgFQJ4AdnQN4F8BYAKBIDMBXAO2ABcBLeK6Ac2Vo2wB5VoBeaADd49ACYA+ABQA6WQH0AXNF7IAHrWRVRMYWOgB+aAG0ARAE14FaIgoRa0TACd4gsTjDRaWHJjCOwKBqOJgC60EpGqPohAJRKCCicOAA+0NSiyGT0VMiiJLgk0EXQjuwUjsw5AO5wSGjekjEA3CREpMSgkDAAQn75hcVsHN48UrT6SqhxtYne0Knpmdm5A0UFxMWbrOxJo5Lmlta29k4ubtAeXtgOfgHsyMHN0AD0z9AA6gAWYPb0MLxVT6aaD0eyieDIGBUeD2FDsC5UdDQCDneBkTxAzwPKh+JHwbD+WjwRyGEirLZDXaoKRPV7QAByMOgVWJAGtsixoEDSgBCcmbUq0crMSkjamSWi0t6M+wsxzsqic5Cg7l8jbFNptEjAJh2Tz8NJaJY5UQtdqi7i9Rw0povN7vNkcrkPZBqi3ILhWsZSj6OxXO3lk4juxq2ukAYWJpTo0Aezkcbp2DUltrt0Ejjmj9jjxJ5QA)

### 💻 Code

```ts
class SomeType {}

function getType<T = void>(..._: T extends void ? ["You must provide a type parameter"] : [T?]): SomeType | undefined
{
    return new SomeType();
}

class Bar
{
    getType<T>(t?: T): SomeType | undefined
    {
        return getType<T>(); // Not working here! <----------------------------------
        return getType<T>(t); // Not working either!

        return getType<T>("You must provide a type parameter"); // Not working either! What is T when it does not meet any side of the ternary operator? 
    }
}

const t = undefined;

getType<Bar>(); // Working here!
getType<Bar>(t); // Working here!

getType(); // Correct error!
getType(t);  // Correct error!
```

### 🙁 Actual behavior
Call of `getType<T>()` function with generic parameter as generic argument does not work the same way as call with specific type `getType<Bar>()`.

### 🙂 Expected behavior
Both cases should work the same way.

**Comment by Codesmith1024** [comment]
I was gonna report this as a separate bug, but I feel like this is probably related to your issue --

[playground code](https://www.typescriptlang.org/play?ts=4.4.3#code/PTAEFEDcEMBsFdoBcCmBnUSD2poDsBPUASwDNQAVEjaNNYgcz2gCNYVMcBVAGlzwAmoPCkgoATqCxIAFhIDuxNCgBQSAgAcOAQTqNmbFAB4KfLgD5QAXkqgUAD1SCMXUAH5+RAFzDREgNwqIKBoSOLEeAygAD7C8LCw1LgaGtDiKHhIsES09Eys7JwhYREMAHRlbmqaOtagunkG7Eah4ZExcQl8raXmgUFgAPIyoPLQxEh8skpSIqAa4lha4tl2MAjI6EUiYuIVZdVaoABCdQAUPe2xePGwAJR2jhkCGBclkQ8e+N6+u4FAA)
```ts
// Evaluates to any if T is assignable to U, and never otherwise
type Assignable<T, U> = T extends U ? any : never;
// string | null is apparently assignable to string..?
type A = Assignable<string | null, string>;

// Oh wait, this one properly evaluates to never...
type B = (string | null) extends (string) ? any : never;
```

Basically, if you hover over `A` (and actually use it in-code), then A has the type `any`, implying that `string | null` can be assigned to type `string`. `B` however, evaluates to `never`, despite doing the exact same check, just without the generic types.

**Comment by MartinJohns** [comment]
@Hookyns There's absolutely no way for TypeScript to figure out what `T` is within your method, both first and second option could be valid, but there's just not enough information to figure it out.

**Comment by fatcerberus** [comment]
To elaborate on the above: My understanding is, when conditional types are instantiated with a generic type parameter, their evaluation is deferred and assignability is determined by an internal check `isTypeIdenticalTo`.  This is a much stricter check than the usual assignability rules.  See my comment here:

https://github.com/microsoft/TypeScript/issues/27024#issuecomment-510924206

This particular check can only be satisfied by another conditional type.  `T` is not a conditional type.

**Comment by Hookyns** [comment]
> @Hookyns There's absolutely no way for TypeScript to figure out what `T` is within your method, both first and second option could be valid, but there's just not enough information to figure it out.

@MartinJohns Even if I use constraint, so it cannot be void (so false branch should be used), it does not work.
_And `t` is not the point._

So @fatcerberus seems be right. 
But this behavior is not intended and my code should be compiled without errors, right? Cuz it is valid code IMHO.

Updated example which is a little more complex but closer to my case. [Playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDKD2BbApgFQJ4AdnQLACgBvA6U6TAVwCMQBLYaAJ2TABN4A7EdaYAF3iMAXNABiFDv1qcA3ARJlgnCH0YV+ggBQo+bMLpGFeA4WIlTOAGmgB6AFTQAdM+fQ7N6AF8AlNGL4yQOg+AAtaCEcNRmgAXmgdPV1IkzkAsk8CDPwCezsSB3FJPmkOaAB3WlDoABEwRgBraABZMABzegBCaAAlZD4KRg4YBBQMbGgqSGRWaE5oVuQOZEZ6cjqwHWXrULAOepgBYIgAWmYAMxBkCw5HfOgAAQhkHBC+PkwIIRsbdtDqSKQNgAEvB4PV0EMbHwTudLtd8jYCGdzMU5gs+GNkAAeVCxaAAN3gtFYAD5NC4APoiXHIAAefEWrBghOJ0AA-NAANoAIgAmvAKPEKCpyIx4PjiTgwMEsDhMOtNoxuQBdaAiTnK7wiEZoWW2DxY45G40mgj+QLMfqDaBLMpwJC67CaIxREQAeSoACsrnxrPZoLsZohBDhaBwzoJEPoSm4bD5UlkCKBIMNlhLgMgAApiiWsZZmhSkdGwNP0bGoarIbAcPOSHh0hk1mAASRLjHTyDJfDZ2odmK10ArVcZi2APG+0ENJunhb8s8CfDxi7ZHPRmJxlertbHZO8Mn10AAcvBF2VBPUw61oCFlsgulPp8b52RLQNSr0Lj6AUNVOo+Jo+GSQRrA1PdZyyRN8DDBlGDOMAM2gVtSwQgs0lICN4E0AcWVYBN5HwZMoHtFA2w7aBaEQTBLhQDhoUQ0iyzwIhZyUH81CiLC-CyQIMM4wgIPw1iRQgTA8VtOBkKzHNJUYLDUiTZRFwgB0GIQuIRMcYtJKxHVVM7OSCGUkjJKcXiwOyfA11lHSVMk3d9wnAB1c9L2vW970fU1LL6TE5IPABhQRmH4aBljFRgOiAA)

```typescript
class SomeType 
{
    public readonly ctor: Function;
    constructor(...) { ... }
}

// https://github.com/Hookyns/ts-reflection
function getType<T = void>(..._: T extends void ? ["You must provide a type parameter"] : []): SomeType // <----------
{
    return new SomeType({ ctor: Object, /* and more information */});
}

class ServiceProvider
{
    getService<TDependency extends IService>(t?: SomeType): TDependency // <--- even with constraint; now it cannot be void, never
    {
        t = t ?? getType<TDependency>(); // Not working here! <----------
        return Reflect.construct(t.ctor, []);
    }
}

interface IService 
{
    foo(): void;
}

class SomeService implements IService 
{
    constructor() {}
    foo() {}
}

const sp = new ServiceProvider();

const someService = sp.getService<SomeService>();
someService .foo();

getType<SomeService>(); // Working here! <----------
getType(); // Correct error, because required generic parameter is missing!
```

And a minimal example:
```typescript
function foo<T = void>(arg: T extends void ? true : false): T | undefined 
{
    return undefined;
}

function callFoo<T extends boolean>()
{
    foo<T>(true); // error, boolean is not assignable...
    foo<T>(false); // same error
}

callFoo<true>();
```

**Comment by Hookyns** [comment]
> It doesn't work here, because TypeScript **doesn't know** what type you're having here. `TDependency` is not an actual type, it's a placeholder for whatever type is passed along. Without knowing the **actual type** it can't resolve the conditional type. It can't verify whether `TDependency extends void`, because it's not known what `TDependency` is. 

Thank You @MartinJohns for explanation what **generic** means. :expressionless:

> And constraints are not considered for this.

So it sounds like TypeScript can and should do that, but TypeScript don't do that. What is the definition of bug, please? 

Doesn't matter if it is difficult to implement or not. Is it valid? Should it work? Yes! Or no? Tell me it is a bug and I can invest my time to try to figure it out and maybe make a PR. Just don't say it is working like a charm and it is intended behavior because You don't want to implement it. Just say OK, it is a bug but it will be almost impossible to implement that. Then I will be happy. 

And what about this? If TypeScript can report error for `const x: T = undefined;` it should be able to use the information elsewhere.
```typescript
function getType<T = undefined>(arg: T extends undefined ? true : false)
{
}

function callFoo<T extends boolean>()
{
    const x: T = undefined; // Error, 'T' is not assignable to type T. Interesting.. TypeScript know what T is, that it cannot be undefined.
    getType<T>(false); // Ohh.. error... it may be undefined.. Sounds like BUG for me.
}
```

**Comment by fatcerberus** [comment]
It’s not a bug.  TS can’t resolve the conditional type because it doesn’t know what the type *is* inside the generic function.  The caller has to provide it.

**Comment by tjjfvi** [comment]
```ts
function foo<T = void>(arg: T extends void ? true : false): T | undefined 
{
    return undefined;
}

function callFoo<T extends boolean>()
{
    foo<T>(true); // error
    foo<T>(false); // hypothetical universe where this doesn't error
}

callFoo<never>() // ok; never extends boolean

foo<never>(false) // this errors because never extends void
                  // note that this didn't error in the `callFoo` invocation
```

**Comment by MartinJohns** [comment]
> Sorry for my sarcasm, but it seems like You are trying to throw everything, no matter what, just to, idk, hide a bug.

I'm not part of the TypeScript team, and I'm not trying to hide anything. I was merely trying to explain you why it's not a bug and you're mistaken.

**Comment by m-rutter** [comment]
@Hookyns  It feels like you are expecting modus tollens to apply to the evaluation of the conditional types. If the value argument is the literal `false`, then the type level expression `T extends undefined` must be false, which makes some kind of sense and probably is intuitive if you come from a propositional logic background. 

However, this just isn't how conditional types are evaluated, changing this should be a very signficant breaking change, and changing this would introduce signficant structure comparsion checks that would create an performance overhead that would make the feature unsuable.

**Comment by fatcerberus** [comment]
> Everyone says that TypeScript doesn't know what T is

To be fair, it would be more accurate to say that the evaluation of the conditional type is *deferred*, because actually evaluating such things in a generic context requires higher-level reasoning than the compiler is capable of.  Basically, we know that the code inside the body of `foo` must work for *any* valid instantiation of `T` - [parametricity](https://en.wikipedia.org/wiki/Parametricity) - literally an infinite number of types, as opposed to one single type that you know upfront.  In a few cases the compiler can--and will--pretend the type parameter is equal to its constraint, but in general this causes more problems than it solves due to variance concerns.  Anyway, while a conditional type is "deferred" in this manner, the compiler doesn't reason about its branches individually - it sees a single atomic thing that it can only do an identity check on.  Hence the behavior you see here.

And as @m-rutter pointed out, TS doesn't do modus tollens inferences.  That kind of reasoning would require negated types - see https://github.com/microsoft/TypeScript/pull/29317.  The compiler can do inferences of the form `(T => U) & T => U` all day long, but it pretty much never does `(T => U) & ~U => ~T`, because it has no way to represent `~T` internally.
